File: routes/news.py
from __future__ import annotations

import re
import shutil
from datetime import datetime
from pathlib import Path

from fastapi import APIRouter, Depends, File, HTTPException, Query, UploadFile
from fastapi.responses import FileResponse
from pydantic import BaseModel
from sqlalchemy.orm import Session

from config import settings
from data.db_session import get_db
from data.news import NewsMedia, NewsPoll, NewsPollOption, NewsPollVote, NewsPost
from data.users import User
from utils.auth_deps import require_kb_editor, require_user
from utils.htmlsanitize import sanitize_html
from utils.logger import logger

router = APIRouter(prefix="/api/news", tags=["news"])

NEWS_DIR = settings.docs_dir / "news"

_IMAGE_EXT = {".png", ".jpg", ".jpeg", ".webp", ".gif", ".bmp"}
_FILE_EXT = {
    ".pdf", ".docx", ".doc", ".txt", ".md", ".rst", ".csv",
    ".xlsx", ".xlsm", ".xls", ".pptx", ".ppt", ".rtf", ".odt", ".ods", ".zip",
}


def _unique_path(target_dir: Path, filename: str) -> Path:
    base, ext = Path(filename).stem, Path(filename).suffix
    candidate = target_dir / filename
    i = 1
    while candidate.exists():
        candidate = target_dir / f"{base} ({i}){ext}"
        i += 1
    return candidate


def _author_name(db: Session, author_id: int | None) -> str:
    if not author_id:
        return "—"
    u = db.get(User, author_id)
    return u.full_name if u else "—"


_IMG_SRC_RE = re.compile(r'<img[^>]+src="([^"]+)"', re.IGNORECASE)
_TAG_RE = re.compile(r"<[^>]+>")


def _preview(body_html: str) -> tuple[str | None, str]:
    """Первая картинка (для обложки) + текстовая выжимка (для превью в ленте)."""
    m = _IMG_SRC_RE.search(body_html or "")
    img = m.group(1) if m else None
    text = _TAG_RE.sub(" ", body_html or "")
    text = re.sub(r"\s+", " ", text).strip()
    excerpt = text[:200] + ("…" if len(text) > 200 else "")
    return img, excerpt


def _post_dict(db: Session, p: NewsPost, user: User) -> dict:
    preview_image, excerpt = _preview(p.body_html)
    return {
        "id": p.id,
        "title": p.title,
        "body_html": p.body_html,
        "attachments": p.attachments or [],
        "preview_image": preview_image,
        "excerpt": excerpt,
        "poll": _poll_edit_dict(db, p.id),  # для префилла редактора
        "author": _author_name(db, p.author_id),
        "is_pinned": p.is_pinned,
        "created_at": p.created_at.isoformat() if p.created_at else None,
        "updated_at": p.updated_at.isoformat() if p.updated_at else None,
    }


# ─────────────────────────── чтение (все сотрудники) ───────────────────────────


@router.get("")
async def list_news(user: User = Depends(require_user), db: Session = Depends(get_db)):
    posts = (
        db.query(NewsPost)
        .filter(NewsPost.is_published.is_(True))
        .order_by(NewsPost.is_pinned.desc(), NewsPost.created_at.desc())
        .all()
    )
    return {
        "success": True,
        "can_edit": bool(user.is_admin or user.is_kb_editor),
        "items": [_post_dict(db, p, user) for p in posts],
    }


@router.get("/media/{media_id}")
async def get_media(
    media_id: int,
    download: bool = Query(default=False),
    as_: str | None = Query(default=None, alias="as"),
    user: User = Depends(require_user),
    db: Session = Depends(get_db),
):
    m = db.get(NewsMedia, media_id)
    if not m:
        raise HTTPException(404, "Файл не найден")
    path = Path(m.stored_path)
    try:
        path = path.resolve()
        path.relative_to(settings.docs_dir.resolve())
    except (ValueError, OSError):
        raise HTTPException(403, "Доступ к файлу запрещён")
    if not path.exists() or not path.is_file():
        raise HTTPException(404, "Файл отсутствует на диске")
    if as_ == "pdf":
        from utils.file_preview import preview_pdf_response

        resp = preview_pdf_response(path)
        if resp is not None:
            return resp
    # Картинки и pdf показываем прямо в браузере; прочее — по ?download=1 качаем.
    inline = not download
    return FileResponse(
        path,
        filename=m.original_name or path.name,
        media_type=m.mime_type or "application/octet-stream",
        content_disposition_type="inline" if inline else "attachment",
    )


# ─────────────────────── загрузка/создание (редакторы БЗ) ──────────────────────


@router.post("/upload")
async def upload_media(
    file: UploadFile = File(...),
    user: User = Depends(require_kb_editor),
    db: Session = Depends(get_db),
):
    """Загрузка картинки или документа для новости. Возвращает media-запись с url —
    редактор вставляет <img> (для картинок) или чип-вложение (для документов)."""
    ext = Path(file.filename or "").suffix.lower()
    is_image = ext in _IMAGE_EXT
    if not (is_image or ext in _FILE_EXT):
        raise HTTPException(400, f"Неподдерживаемый формат: {ext or '—'}")

    NEWS_DIR.mkdir(parents=True, exist_ok=True)
    target = _unique_path(NEWS_DIR, Path(file.filename or "file").name)
    with target.open("wb") as f:
        shutil.copyfileobj(file.file, f)
    size = target.stat().st_size

    m = NewsMedia(
        original_name=Path(file.filename or target.name).name,
        stored_path=str(target),
        mime_type=file.content_type,
        size=size,
        is_image=is_image,
        uploaded_by=user.id,
    )
    db.add(m)
    db.commit()
    db.refresh(m)

    return {
        "success": True,
        "media": {
            "id": m.id,
            "name": m.original_name,
            "size": m.size,
            "is_image": m.is_image,
            "url": f"/api/news/media/{m.id}",
        },
    }


class PollPayload(BaseModel):
    question: str = ""
    description: str | None = None
    allow_multiple: bool = False
    show_voters: bool = False
    options: list[str] = []


class PostPayload(BaseModel):
    title: str = ""
    body_html: str = ""
    attachments: list[dict] | None = None
    is_pinned: bool = False
    poll: PollPayload | None = None


def _save_poll(db: Session, post: NewsPost, pp: PollPayload | None) -> None:
    """Создаёт/заменяет голосование поста. None или <2 вариантов — удаляет.
    Замена сбрасывает прежние голоса (каскадом)."""
    existing = db.query(NewsPoll).filter(NewsPoll.post_id == post.id).first()
    opts = [o.strip() for o in (pp.options if pp else []) if o and o.strip()]
    if not pp or not pp.question.strip() or len(opts) < 2:
        if existing:
            db.delete(existing)
        return
    q = pp.question.strip()[:300]
    desc = (pp.description or "").strip() or None
    trimmed = [o[:300] for o in opts[:12]]
    if existing:
        cur_opts = [
            o.text for o in db.query(NewsPollOption)
            .filter(NewsPollOption.poll_id == existing.id)
            .order_by(NewsPollOption.position).all()
        ]
        unchanged = (
            existing.question == q and existing.description == desc
            and existing.allow_multiple == bool(pp.allow_multiple)
            and existing.show_voters == bool(pp.show_voters)
            and cur_opts == trimmed
        )
        if unchanged:
            return  # опрос не изменился — сохраняем голоса
        db.delete(existing)
        db.flush()
    poll = NewsPoll(
        post_id=post.id, question=q, description=desc,
        allow_multiple=bool(pp.allow_multiple), show_voters=bool(pp.show_voters),
    )
    db.add(poll)
    db.flush()
    for i, text in enumerate(trimmed):
        db.add(NewsPollOption(poll_id=poll.id, text=text, position=i))


def _poll_state(db: Session, poll: NewsPoll, user: User) -> dict:
    options = (
        db.query(NewsPollOption)
        .filter(NewsPollOption.poll_id == poll.id)
        .order_by(NewsPollOption.position)
        .all()
    )
    votes = db.query(NewsPollVote).filter(NewsPollVote.poll_id == poll.id).all()
    by_opt: dict[int, list[int]] = {}
    for v in votes:
        by_opt.setdefault(v.option_id, []).append(v.user_id)
    total = len(votes)
    mine = {v.option_id for v in votes if v.user_id == user.id}

    def voters(uids: list[int]) -> list[dict]:
        if not poll.show_voters:
            return []
        out = []
        for uid in uids:
            u = db.get(User, uid)
            if u:
                out.append({"name": u.full_name, "initials": u.initials})
        return out

    return {
        "id": poll.id,
        "question": poll.question,
        "description": poll.description,
        "allow_multiple": poll.allow_multiple,
        "show_voters": poll.show_voters,
        "total_votes": total,
        "options": [
            {
                "id": o.id,
                "text": o.text,
                "votes": len(by_opt.get(o.id, [])),
                "mine": o.id in mine,
                "voters": voters(by_opt.get(o.id, [])),
            }
            for o in options
        ],
    }


def _poll_edit_dict(db: Session, post_id: int) -> dict | None:
    """Данные голосования для префилла редактора (без счётчиков)."""
    poll = db.query(NewsPoll).filter(NewsPoll.post_id == post_id).first()
    if not poll:
        return None
    opts = (
        db.query(NewsPollOption)
        .filter(NewsPollOption.poll_id == poll.id)
        .order_by(NewsPollOption.position)
        .all()
    )
    return {
        "question": poll.question,
        "description": poll.description or "",
        "allow_multiple": poll.allow_multiple,
        "show_voters": poll.show_voters,
        "options": [o.text for o in opts],
    }


def _resolve_attachments(db: Session, raw: list[dict] | None) -> list[dict]:
    """По списку от клиента берём media_id, перепроверяем существование в БД и
    собираем достоверные метаданные (имя/размер/url с сервера, не с клиента)."""
    out: list[dict] = []
    for item in raw or []:
        mid = item.get("media_id") if isinstance(item, dict) else None
        if not mid:
            continue
        m = db.get(NewsMedia, int(mid))
        if not m:
            continue
        out.append({
            "media_id": m.id,
            "name": m.original_name,
            "size": m.size,
            "is_image": m.is_image,
            "url": f"/api/news/media/{m.id}",
        })
    return out


def _bind_media(db: Session, post: NewsPost) -> None:
    """Привязываем к посту media из вложений и встроенные в текст картинки/документы —
    чтобы при удалении поста файлы удалились каскадом."""
    ids: set[int] = {int(a["media_id"]) for a in (post.attachments or []) if a.get("media_id")}
    ids.update(int(x) for x in re.findall(r"/api/news/media/(\d+)", post.body_html or ""))
    for mid in ids:
        m = db.get(NewsMedia, mid)
        if m and m.post_id is None:
            m.post_id = post.id


@router.post("")
async def create_post(
    payload: PostPayload,
    user: User = Depends(require_kb_editor),
    db: Session = Depends(get_db),
):
    title = (payload.title or "").strip()
    body = sanitize_html(payload.body_html or "")
    attachments = _resolve_attachments(db, payload.attachments)
    if not title and not body and not attachments:
        raise HTTPException(400, "Пустая новость")

    post = NewsPost(
        title=title or "Без заголовка",
        body_html=body,
        attachments=attachments or None,
        author_id=user.id,
        is_pinned=bool(payload.is_pinned),
    )
    db.add(post)
    db.commit()
    db.refresh(post)
    _bind_media(db, post)
    _save_poll(db, post, payload.poll)
    db.commit()
    logger.info("[NEWS] пост {} создан пользователем {}", post.id, user.id)
    return {"success": True, "post": _post_dict(db, post, user)}


@router.patch("/{post_id}")
async def update_post(
    post_id: int,
    payload: PostPayload,
    user: User = Depends(require_kb_editor),
    db: Session = Depends(get_db),
):
    post = db.get(NewsPost, post_id)
    if not post:
        raise HTTPException(404, "Новость не найдена")
    post.title = (payload.title or "").strip() or "Без заголовка"
    post.body_html = sanitize_html(payload.body_html or "")
    post.attachments = _resolve_attachments(db, payload.attachments) or None
    post.is_pinned = bool(payload.is_pinned)
    post.updated_at = datetime.now()
    _bind_media(db, post)
    _save_poll(db, post, payload.poll)
    db.commit()
    db.refresh(post)
    logger.info("[NEWS] пост {} обновлён пользователем {}", post.id, user.id)
    return {"success": True, "post": _post_dict(db, post, user)}


@router.delete("/{post_id}")
async def delete_post(
    post_id: int,
    user: User = Depends(require_kb_editor),
    db: Session = Depends(get_db),
):
    post = db.get(NewsPost, post_id)
    if not post:
        raise HTTPException(404, "Новость не найдена")
    # Удаляем файлы привязанных media с диска (каскад в БД снесёт записи).
    media = db.query(NewsMedia).filter(NewsMedia.post_id == post_id).all()
    for m in media:
        try:
            p = Path(m.stored_path).resolve()
            p.relative_to(settings.docs_dir.resolve())
            if p.exists():
                p.unlink()
        except (ValueError, OSError):
            pass
    db.delete(post)
    db.commit()
    logger.info("[NEWS] пост {} удалён пользователем {}", post_id, user.id)
    return {"success": True}


# ─────────────────────────── голосование ───────────────────────────


@router.get("/{post_id}/poll")
async def get_poll(
    post_id: int,
    user: User = Depends(require_user),
    db: Session = Depends(get_db),
):
    poll = db.query(NewsPoll).filter(NewsPoll.post_id == post_id).first()
    if not poll:
        return {"success": True, "poll": None}
    return {"success": True, "poll": _poll_state(db, poll, user)}


class VotePayload(BaseModel):
    option_id: int


@router.post("/poll/vote")
async def vote_poll(
    payload: VotePayload,
    user: User = Depends(require_user),
    db: Session = Depends(get_db),
):
    opt = db.get(NewsPollOption, payload.option_id)
    if not opt:
        raise HTTPException(404, "Вариант не найден")
    poll = db.get(NewsPoll, opt.poll_id)
    if not poll:
        raise HTTPException(404, "Голосование не найдено")

    already = (
        db.query(NewsPollVote)
        .filter(
            NewsPollVote.poll_id == poll.id,
            NewsPollVote.option_id == opt.id,
            NewsPollVote.user_id == user.id,
        )
        .first()
    )
    if already:
        db.delete(already)  # повторный клик — снять голос
    else:
        if not poll.allow_multiple:
            # одиночный выбор — снимаем прежние голоса пользователя в этом опросе
            db.query(NewsPollVote).filter(
                NewsPollVote.poll_id == poll.id, NewsPollVote.user_id == user.id
            ).delete()
        db.add(NewsPollVote(poll_id=poll.id, option_id=opt.id, user_id=user.id))
    db.commit()
    return {"success": True, "poll": _poll_state(db, poll, user)}

File: app/lib/news.ts
import 'server-only';
import path from 'node:path';
import { NextResponse } from 'next/server';
import { isoUtc, prisma } from './db';

// Общая логика домена «Новости» — порт вспомогательных функций
// backend/routes/news.py (_post_dict, _preview, _poll_edit_dict, _poll_state,
// _author_name, _resolve_attachments, _bind_media, _save_poll).
// Вынесена сюда, потому что её делят между собой пять route-handler'ов.

// ── пути к файлам ──────────────────────────────────────────────────────────
// Файлы новостей лежат там же, где их кладёт FastAPI: backend/docs/news.
// Иначе старые и новые вложения окажутся в разных каталогах и часть
// медиа перестанет открываться. process.cwd() в Next — корень app/.

const BACKEND_DIR = process.env.BACKEND_DIR
  ? path.resolve(process.env.BACKEND_DIR)
  : path.resolve(process.cwd(), '..', 'backend');

export const DOCS_DIR = path.join(BACKEND_DIR, 'docs');
export const NEWS_DIR = path.join(DOCS_DIR, 'news');

/**
 * Абсолютный путь к файлу, если он внутри docs_dir (аналог resolve() +
 * relative_to() в Python). Иначе null — наружу отдавать/удалять нельзя.
 * На Windows сравниваем без учёта регистра, как это делает pathlib.
 *
 * В БД пути хранятся относительно docs_dir — так переезд или переименование
 * каталога проекта не ломает ссылки. Абсолютные значения (записи старых
 * версий) тоже принимаются.
 */
export function resolveInsideDocs(stored: string): string | null {
  if (!stored) return null;
  if (/^[a-z][a-z0-9+.-]*:\/\//i.test(stored)) return null; // URL — не файл
  const target = path.isAbsolute(stored) ? path.resolve(stored) : path.resolve(DOCS_DIR, stored);
  const root = path.resolve(DOCS_DIR);
  const norm = (s: string) => (process.platform === 'win32' ? s.toLowerCase() : s);
  const a = norm(target);
  const b = norm(root);
  if (a === b) return target;
  return a.startsWith(b.endsWith(path.sep) ? b : b + path.sep) ? target : null;
}

/** Абсолютный fs-путь по хранимому значению (без проверки принадлежности). */
export function fromDocsPath(stored: string): string {
  return path.isAbsolute(stored) ? stored : path.resolve(DOCS_DIR, stored);
}

/**
 * Вид пути для записи в БД: относительный к docs_dir. Значение вне docs_dir
 * (в том числе URL или уже относительный путь) возвращается как есть.
 */
export function toDocsPath(p: string): string {
  if (!p || !path.isAbsolute(p)) return p;
  const inside = resolveInsideDocs(p);
  return inside ? path.relative(path.resolve(DOCS_DIR), inside) : p;
}

// ── строковые утилиты в стиле pathlib/pydantic ─────────────────────────────

/** Обрезка по кодовым точкам — Python режет строки так же, а не по UTF-16. */
export function cut(s: string, n: number): string {
  const chars = Array.from(s);
  return chars.length <= n ? s : chars.slice(0, n).join('');
}

/** Имя файла без каталогов (Path(...).name, оба вида разделителей). */
export function baseName(filename: string): string {
  const parts = filename.split(/[\\/]/);
  return parts[parts.length - 1] ?? '';
}

/** Расширение по правилам Path.suffix: у «.gitignore» и «file.» его нет. */
export function suffixOf(name: string): string {
  const i = name.lastIndexOf('.');
  return i > 0 && i < name.length - 1 ? name.slice(i) : '';
}

/** Имя без расширения (Path.stem). */
export function stemOf(name: string): string {
  const suffix = suffixOf(name);
  return suffix ? name.slice(0, name.length - suffix.length) : name;
}

// ── ошибки в формате FastAPI ───────────────────────────────────────────────

/** RequestValidationError FastAPI: 422 со списком detail. */
export function validationError(
  loc: (string | number)[],
  type: string,
  msg: string,
  input: unknown
): NextResponse {
  return NextResponse.json({ detail: [{ type, loc, msg, input }] }, { status: 422 });
}

/** Числовой параметр пути: нечисловой сегмент FastAPI отвергает с 422. */
export function parseIntParam(
  raw: string,
  name: string
): { value: number } | { response: NextResponse } {
  if (!/^[+-]?\d+$/.test(raw)) {
    return {
      response: validationError(
        ['path', name],
        'int_parsing',
        'Input should be a valid integer, unable to parse string as an integer',
        raw
      ),
    };
  }
  return { value: Number.parseInt(raw, 10) };
}

// ── типы ответов ───────────────────────────────────────────────────────────

// Именно type, а не interface: только у псевдонима типа есть неявная
// индексная сигнатура, без которой массив не подходит под Prisma.InputJsonValue.
export type AttachmentDict = {
  media_id: number;
  name: string;
  size: number;
  is_image: boolean;
  url: string;
};

export interface PollDraft {
  question: string;
  description: string;
  allow_multiple: boolean;
  show_voters: boolean;
  options: string[];
}

export interface PostDict {
  id: number;
  title: string;
  body_html: string;
  attachments: unknown[];
  preview_image: string | null;
  excerpt: string;
  poll: PollDraft | null;
  author: string;
  is_pinned: boolean;
  created_at: string | null;
  updated_at: string | null;
}

export interface PollVoterDict {
  name: string;
  initials: string;
}

export interface PollOptionDict {
  id: number;
  text: string;
  votes: number;
  mine: boolean;
  voters: PollVoterDict[];
}

export interface PollStateDict {
  id: number;
  question: string;
  description: string | null;
  allow_multiple: boolean;
  show_voters: boolean;
  total_votes: number;
  options: PollOptionDict[];
}

export interface PostRow {
  id: number;
  title: string;
  body_html: string;
  attachments: unknown;
  author_id: number | null;
  is_pinned: boolean;
  created_at: Date;
  updated_at: Date | null;
}

// ── превью ─────────────────────────────────────────────────────────────────

const IMG_SRC_RE = /<img[^>]+src="([^"]+)"/i;
const TAG_RE = /<[^>]+>/g;

/** Первая картинка (для обложки) + текстовая выжимка (для превью в ленте). */
export function preview(bodyHtml: string): [string | null, string] {
  const m = IMG_SRC_RE.exec(bodyHtml || '');
  const img = m ? m[1] : null;
  const text = (bodyHtml || '').replace(TAG_RE, ' ').replace(/\s+/g, ' ').trim();
  const chars = Array.from(text);
  const excerpt = chars.slice(0, 200).join('') + (chars.length > 200 ? '…' : '');
  return [img, excerpt];
}

// ── сборка ответа по посту ─────────────────────────────────────────────────

function fullNameOf(u: { surname: string; name: string; patronymic: string | null }): string {
  return [u.surname, u.name, u.patronymic || ''].filter(Boolean).join(' ').trim();
}

function initialsOf(u: { surname: string; name: string }): string {
  const first = (u.name || '').trim().slice(0, 1);
  const last = (u.surname || '').trim().slice(0, 1);
  return (last + first).toUpperCase();
}

/**
 * _post_dict для набора постов сразу: авторы и голосования подтягиваются
 * пакетно (в Python на каждый пост шёл отдельный запрос — результат тот же).
 */
export async function postDicts(posts: PostRow[]): Promise<PostDict[]> {
  if (!posts.length) return [];

  const authorIds = [...new Set(posts.map((p) => p.author_id).filter((v): v is number => !!v))];
  const authors = authorIds.length
    ? await prisma.users.findMany({
        where: { id: { in: authorIds } },
        select: { id: true, surname: true, name: true, patronymic: true },
      })
    : [];
  const nameById = new Map<number, string>(authors.map((u) => [u.id, fullNameOf(u)]));

  const polls = await prisma.news_polls.findMany({
    where: { post_id: { in: posts.map((p) => p.id) } },
    orderBy: { id: 'asc' },
    include: { news_poll_options: { orderBy: { position: 'asc' } } },
  });
  const pollByPost = new Map<number, PollDraft>();
  for (const poll of polls) {
    if (pollByPost.has(poll.post_id)) continue; // как .first() в Python — берём один
    pollByPost.set(poll.post_id, {
      question: poll.question,
      description: poll.description || '',
      allow_multiple: poll.allow_multiple,
      show_voters: poll.show_voters,
      options: poll.news_poll_options.map((o) => o.text),
    });
  }

  return posts.map((p) => {
    const [previewImage, excerpt] = preview(p.body_html);
    const attachments = Array.isArray(p.attachments) ? p.attachments : [];
    return {
      id: p.id,
      title: p.title,
      body_html: p.body_html,
      attachments,
      preview_image: previewImage,
      excerpt,
      poll: pollByPost.get(p.id) ?? null, // для префилла редактора
      author: (p.author_id && nameById.get(p.author_id)) || '—',
      is_pinned: p.is_pinned,
      created_at: isoUtc(p.created_at),
      updated_at: isoUtc(p.updated_at),
    };
  });
}

export async function postDict(post: PostRow): Promise<PostDict> {
  const [dict] = await postDicts([post]);
  return dict;
}

// ── вложения и привязка медиа ──────────────────────────────────────────────

/**
 * По списку от клиента берём media_id, перепроверяем существование в БД и
 * собираем достоверные метаданные (имя/размер/url с сервера, не с клиента).
 */
export async function resolveAttachments(raw: unknown): Promise<AttachmentDict[]> {
  if (!Array.isArray(raw) || !raw.length) return [];

  const ids: number[] = [];
  for (const item of raw) {
    if (!item || typeof item !== 'object') continue;
    const value = (item as Record<string, unknown>).media_id;
    const id =
      typeof value === 'number'
        ? Math.trunc(value)
        : typeof value === 'string' && /^[+-]?\d+$/.test(value.trim())
          ? Number.parseInt(value, 10)
          : 0;
    if (!id) continue;
    ids.push(id);
  }
  if (!ids.length) return [];

  const media = await prisma.news_media.findMany({ where: { id: { in: [...new Set(ids)] } } });
  const byId = new Map(media.map((m) => [m.id, m] as const));

  const out: AttachmentDict[] = [];
  for (const id of ids) {
    const m = byId.get(id);
    if (!m) continue;
    out.push({
      media_id: m.id,
      name: m.original_name,
      size: m.size,
      is_image: m.is_image,
      url: `/api/news/media/${m.id}`,
    });
  }
  return out;
}

/**
 * Привязываем к посту media из вложений и встроенные в текст картинки/документы —
 * чтобы при удалении поста файлы удалились каскадом.
 */
export async function bindMedia(
  postId: number,
  attachments: AttachmentDict[],
  bodyHtml: string
): Promise<void> {
  const ids = new Set<number>();
  for (const a of attachments) {
    if (a.media_id) ids.add(a.media_id);
  }
  for (const m of (bodyHtml || '').matchAll(/\/api\/news\/media\/(\d+)/g)) {
    ids.add(Number.parseInt(m[1], 10));
  }
  if (!ids.size) return;
  // post_id: null — привязываем только «ничьи» медиа, как в Python.
  await prisma.news_media.updateMany({
    where: { id: { in: [...ids] }, post_id: null },
    data: { post_id: postId },
  });
}

// ── голосования ────────────────────────────────────────────────────────────

export interface PollPayload {
  question: string;
  description: string | null;
  allow_multiple: boolean;
  show_voters: boolean;
  options: string[];
}

/**
 * Создаёт/заменяет голосование поста. null или <2 вариантов — удаляет.
 * Замена сбрасывает прежние голоса (каскадом).
 */
export async function savePoll(postId: number, pp: PollPayload | null): Promise<void> {
  const existing = await prisma.news_polls.findFirst({
    where: { post_id: postId },
    orderBy: { id: 'asc' },
  });
  const opts = (pp?.options ?? []).map((o) => o.trim()).filter((o) => o);
  if (!pp || !pp.question.trim() || opts.length < 2) {
    if (existing) await prisma.news_polls.delete({ where: { id: existing.id } });
    return;
  }

  const q = cut(pp.question.trim(), 300);
  const desc = (pp.description || '').trim() || null;
  const trimmed = opts.slice(0, 12).map((o) => cut(o, 300));

  if (existing) {
    const curOpts = await prisma.news_poll_options.findMany({
      where: { poll_id: existing.id },
      orderBy: { position: 'asc' },
      select: { text: true },
    });
    const unchanged =
      existing.question === q &&
      existing.description === desc &&
      existing.allow_multiple === pp.allow_multiple &&
      existing.show_voters === pp.show_voters &&
      curOpts.length === trimmed.length &&
      curOpts.every((o, i) => o.text === trimmed[i]);
    if (unchanged) return; // опрос не изменился — сохраняем голоса
    await prisma.news_polls.delete({ where: { id: existing.id } });
  }

  const poll = await prisma.news_polls.create({
    data: {
      post_id: postId,
      question: q,
      description: desc,
      allow_multiple: pp.allow_multiple,
      show_voters: pp.show_voters,
    },
  });
  await prisma.news_poll_options.createMany({
    data: trimmed.map((text, i) => ({ poll_id: poll.id, text, position: i })),
  });
}

export interface PollRow {
  id: number;
  question: string;
  description: string | null;
  allow_multiple: boolean;
  show_voters: boolean;
}

/** Состояние голосования для конкретного пользователя (_poll_state). */
export async function pollState(poll: PollRow, userId: number): Promise<PollStateDict> {
  const options = await prisma.news_poll_options.findMany({
    where: { poll_id: poll.id },
    orderBy: { position: 'asc' },
  });
  const votes = await prisma.news_poll_votes.findMany({
    where: { poll_id: poll.id },
    orderBy: { id: 'asc' },
  });

  const byOpt = new Map<number, number[]>();
  for (const v of votes) {
    const list = byOpt.get(v.option_id);
    if (list) list.push(v.user_id);
    else byOpt.set(v.option_id, [v.user_id]);
  }
  const mine = new Set(votes.filter((v) => v.user_id === userId).map((v) => v.option_id));

  // Имена голосовавших нужны только в открытом опросе.
  const voterById = new Map<number, PollVoterDict>();
  if (poll.show_voters) {
    const uids = [...new Set(votes.map((v) => v.user_id))];
    if (uids.length) {
      const users = await prisma.users.findMany({
        where: { id: { in: uids } },
        select: { id: true, surname: true, name: true, patronymic: true },
      });
      for (const u of users) {
        voterById.set(u.id, { name: fullNameOf(u), initials: initialsOf(u) });
      }
    }
  }

  return {
    id: poll.id,
    question: poll.question,
    description: poll.description,
    allow_multiple: poll.allow_multiple,
    show_voters: poll.show_voters,
    total_votes: votes.length,
    options: options.map((o) => {
      const uids = byOpt.get(o.id) ?? [];
      return {
        id: o.id,
        text: o.text,
        votes: uids.length,
        mine: mine.has(o.id),
        voters: poll.show_voters
          ? uids.map((uid) => voterById.get(uid)).filter((v): v is PollVoterDict => !!v)
          : [],
      };
    }),
  };
}

// ── разбор тела запроса (PostPayload) ──────────────────────────────────────

export interface PostPayload {
  title: string;
  body_html: string;
  attachments: unknown;
  is_pinned: boolean;
  poll: PollPayload | null;
}

function asString(v: unknown, fallback = ''): string {
  return typeof v === 'string' ? v : fallback;
}

function parsePoll(raw: unknown): PollPayload | null {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) return null;
  const o = raw as Record<string, unknown>;
  return {
    question: asString(o.question),
    description: typeof o.description === 'string' ? o.description : null,
    allow_multiple: Boolean(o.allow_multiple),
    show_voters: Boolean(o.show_voters),
    options: Array.isArray(o.options) ? o.options.filter((x): x is string => typeof x === 'string') : [],
  };
}

/** Значения по умолчанию — как у PostPayload/PollPayload (pydantic). */
export function parsePostPayload(raw: unknown): PostPayload {
  const o = raw && typeof raw === 'object' ? (raw as Record<string, unknown>) : {};
  return {
    title: asString(o.title),
    body_html: asString(o.body_html),
    attachments: o.attachments ?? null,
    is_pinned: Boolean(o.is_pinned),
    poll: parsePoll(o.poll),
  };
}
